docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+**Pros**:
+
+* Low overhead tracking provided by the kernel.
+
+
+**Cons**
+
+* Requires updates to default ownership on docker’s internal storage driver directories. We will have to deal with storage driver implementation implementation details in any approach that is not docker native.
+
+* Requires additional node configuration - quota subsystem needs to be setup on the node. This can either be automated or made a requirement for the node.
+
+* Kubelet needs to perform gid management. A range of gids have to allocated to the kubelet for the purposes of quota management. This range must not be used for any other purposes out of band.
+
+* Breaks `docker save` semantics. Since kubernetes assumes immutable images, this is not a blocker. To support quota in docker, we will need user-namespaces along with custom gid mapping for each container. This feature does not exist today.
+
+*Note: Refer to the [Appendix](#appendix) section more real examples on using quota with docker.*
+
+
+3. Enhance docker
+
+	Docker handles the upload and download of image layers. It can embed enough information about each layer. If docker is enhanced to expose this information, we can statically identify space about to be occupied by read-only image layers, even before the image layers are downloaded.
+
+	A new [docker feature](https://github.com/docker/docker/pull/16450) (docker pull --dry-run) is pending review, which outputs the disk space that will be consumed by new images. Once this feature lands, we can perform feasibility checks and reject pods that will consume more disk space that what is current availability on the node.
+
+	We can also attempt to use this feature to determine image layer’s disk usage statically. The metadata files that docker creates for these layers will not be tracked though.
+
+**Pros**
+
+* Works across all storage drivers since docker abstracts the storage drivers.
+
+* Less code to maintain in kubelet.
+
+**Cons**
+
+* Not available today.
+
+* Requires serialized image pulls.
+
+* Metadata files are not tracked.
+
+
+##### Devicemapper
+
+Devicemapper storage driver will setup two volumes, metadata and data, that will be used to store image layers and container writable layer. The volumes can be real devices or loopback. A Pool device is created which uses the underlying volume for real storage.
+
+A new thinly-provisioned volume, based on the pool, will be created for running container’s.
+
+The kernel tracks the usage of the pool device at the block device layer. The usage here includes image layers and container’s writable layers.
+
+Since the kubelet has to track the writable layer usage anyways, we can subtract the aggregated root filesystem usage from the overall pool device usage to get the image layer’s disk usage.
+
```

Comment:
Ah! Got it. Is it possible to get the usage of just the write layer? Isn't
the writable layer backed by a dedicated volume?

On Thu, Nov 12, 2015 at 1:36 PM, Vivek Goyal <notifications@github.com>
wrote:

> In docs/proposals/disk-accounting.md
> <https://github.com/kubernetes/kubernetes/pull/16889#discussion_r44718149>
> :
>
> > +
> > +* Requires serialized image pulls.
> > +
> > +* Metadata files are not tracked.
> > +
> > +
> > +##### Devicemapper
> > +
> > +Devicemapper storage driver will setup two volumes, metadata and data, that will be used to store image layers and container writable layer. The volumes can be real devices or loopback. A Pool device is created which uses the underlying volume for real storage.
> > +
> > +A new thinly-provisioned volume, based on the pool, will be created for running container’s.
> > +
> > +The kernel tracks the usage of the pool device at the block device layer. The usage here includes image layers and container’s writable layers.
> > +
> > +Since the kubelet has to track the writable layer usage anyways, we can subtract the aggregated root filesystem usage from the overall pool device usage to get the image layer’s disk usage.
> > +
>
> @vishh <https://github.com/vishh> I was thinking of following scenario.
> Say a container is started with image 'foo' which container a file 'bar' of
> say 1G. container now overwrites this
> file 'bar' with 1G of data, then COW will happen in thin pool, and fresh
> 1G of blocks will be allocated. But if we do 'stat -f /' then container
> total usage will not vary and subtracting image size will show size of
> writable layer as 0.
>
> IOW, container wrote 1G of data but still we calculated writable layer
> size as 0. And that's what I was trying to point to if we go with the
> method of subtracting image size from container size.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16889/files#r44718149>.
>